Thermotron3800

jcarbelbide opened the pull request:
This driver has very basic functionality. The following features have been implemented and tested with an oven in our lab.  
[id, init, run, stop, read setpoint, set setpoint, read temperature, read mode]
 
I noticed that wait time is needed after writing anything to the oven (run/stop command, setpoint command). The wait time required appeared to be around 800+ms. I suggested to wait for 1000ms in the docstrings. Queries do not seem to require this wait time. A wait time of 3000ms is required after running the init command. I was not sure whether to embed this wait time in the driver itself, or leave it up to the user. 

My oven would sometimes need to be manually reset (go to the physical oven and restart it) if I did not have adequate wait time.

msmttchr reviewed (changes requested):
@jcarbelbide thanks for your contribution. I made few comments on your implementation.
If not done, I would suggest also a quick look at [documentation](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html), just to make sure nothing is missed.
I would say that also You should add yourself among the authors (files [`authors.rst`](https://github.com/pymeasure/pymeasure/blob/dfca962f2f90ee304fd765b5e90c2bb43e5e7551/docs/about/authors.rst) and [`AUTHORS.txt`](https://github.com/pymeasure/pymeasure/blob/dfca962f2f90ee304fd765b5e90c2bb43e5e7551/AUTHORS.txt)), amend file [`__init__.py`](https://github.com/pymeasure/pymeasure/blob/dfca962f2f90ee304fd765b5e90c2bb43e5e7551/pymeasure/instruments/__init__.py) and [`index.rst`](https://github.com/pymeasure/pymeasure/blob/dfca962f2f90ee304fd765b5e90c2bb43e5e7551/docs/api/instruments/index.rst)

jcarbelbide commented:
Thank you so much for the thorough review, msmttchr! I'll start to implement your suggestions and notify you when everything has been updated.

jcarbelbide commented on pymeasure/instruments/thermotron/thermotron3800.py:
```diff
@@ -0,0 +1,129 @@
+
+    mode = Instrument.measurement(
+        "MODE??", """ Gets the operating mode of the oven.
+        
```
Yes, I think you're right. I checked the documentation, and it says the query is ```MODE?```. I have removed the extra ```?```.

Thanks!

jcarbelbide commented on pymeasure/instruments/thermotron/thermotron3800.py:
```diff
@@ -0,0 +1,129 @@
+        of the oven in Celsius. This property can be set. 
+        
+        Insert wait time after this command. This wait time should be >1000ms for consistent results.
+        Failing to wait for an adequate time may cause errors in subsequent oven commands.
```
Thank you! I really like this solution. I have implemented it into my driver, and updated the comments for the other commands. 

I am a bit unsure what to do about the ```initialize_oven``` command, though. The manufacturer recommends a 3 second wait time after sending this commands. This is a long time to block the user, while they could be calling other instrument initialization functions and take care of the wait themselves. If there are multiple ovens to initialize, it would take 6 seconds, as opposed to just 3 seconds if the user has control of the wait (send both init commands and then insert the wait after). Should I incorporate the wait time into the init function, or allow the user to have control? Or would it be better to incorporate it into the function, but have an optional argument to disable the wait?

jcarbelbide replied:
Hi msmttchr, I added an optional wait argument to the initialize_oven function. The wait time from the initialize_oven function is 2 seconds, for a total of 3 when calling the initalize_oven command (2 seconds from the function + 1 second built into the write function).

jcarbelbide commented on pymeasure/instruments/thermotron/thermotron3800.py:
```diff
@@ -0,0 +1,129 @@
+            64: "Unused (Error)",
+            128: "Calibration mode"
+        }
+
```
Actually, I don't think a combined mode is possible, so I'll remove the comment about the "combined mode". 

Regarding the Python Enum, do you mean something like the following?:
```Python
class Mode(Enum):
    field1 = "string"
    ...
```
I chose the map because I could easily map the integer I get back from the oven to a string. I'm not sure how to do that with the Enum. Can you please help explain what you mean?

bilderbuchi replied:
Something like this?
``` 
class Mode(IntFlag):
    PROGRAM_MODE = 1
    EDIT_MODE = 2
...  #(caps are only convention)
```
See [docs](https://docs.python.org/3/library/enum.html#intflag). Depending on your needs, an `IntEnum` could also be sufficient (if you don't need bitwise combinations - maybe this is the "combined mode" you're referring to?)

bilderbuchi replied:
The enum member you get back is also readable in the console, the name should be sufficiently understandable without having separate strings, and you only need to return one object instead of a tuple from `self.mode`.

jcarbelbide replied:
Hi bilderbuchi, thanks so much for the suggestions! I've added a mode enum to the class, and also added support for that "combined mode" that I mentioned earlier. I used the IntFlag enum for this. I'm not actually sure if it is possible for the oven to be in multiple modes at the same time (which is what I meant by combined mode), but if it ever does happen, the __translate_mode function will now return an enum that has all of the modes OR'd together. For example, it could return:
```Python
Thermotron3800Mode.CALIBRATION_MODE|MANUAL_MODE
```
if the oven is in calibration mode and manual mode. If it is in an unknown mode (mode = 0 or greater than 255), then __translate_mode will return 
```Python
Thermotron3800Mode.UNKNOWN_MODE
```

Please let me know if you think any of this behavior could be tweaked to be better, or if it could be implemented in a better way. I appreciate your feedback. Thanks!

jcarbelbide replied:
[Link to commit](https://github.com/pymeasure/pymeasure/pull/606/commits/7152c11cb1019ded906a939856dfc11831c0d7fe)

Sorry about the linter issues. I believe I have fixed them in future commits.

bilderbuchi replied:
I don't have time right now to review that in detail, but iirc `IntFlag` should be able to do the "or" combination as-is, without any custom code needed. Did you try that before?

E.g., with the `Mode` snippet from above:
```
class Mode(IntFlag):
    PROGRAM_MODE = 1
    EDIT_MODE = 2

Mode(3)  # prints <Mode.EDIT_MODE|PROGRAM_MODE: 3>
```

msmttchr replied:
Addressed by committ 86b2d2a99c9743ebc422ce616ec431533857b7ef as suggested by @bilderbuchi

bilderbuchi commented on pymeasure/instruments/thermotron/thermotron3800.py:
```diff
@@ -0,0 +1,129 @@
+        
+        :return: Tuple(String, int)
+        """,
+        get_process=lambda mode: Thermotron3800.__translate_mode(mode)
```
`get_process=Thermotron3800.__translate_mode` should be sufficient I think (instead of the one-arg pass-through lambda).

jcarbelbide replied:
Thanks for the suggestion! I tried removing the one-arg pass-through lambda, but I got an error (Thermotron3800 not defined) when I ran it. I tried playing around with different schemes, but I couldn't figure out how to get it to work without the lambda function.

bilderbuchi replied:
Ah, yeah. For that to work `__translate_mode` would have to be a standalone function, not a class method, because at import time at this line, `Thermotron3800` does not yet exist (because the class is not fully defined, yet). Sorry, I forgot about that detail.

jcarbelbide replied:
Got it, thanks for explaining! I wasn't sure why, so that was helpful :)

jcarbelbide commented on pymeasure/instruments/thermotron/thermotron3800.py:
```diff
@@ -0,0 +1,129 @@
+        super().__init__(
+            resourceName,
+            "Thermotron 3800",
+            **kwargs
```
Done, thank you!

jcarbelbide commented:
I believe I have addressed all of the requested changes. I have retested all of the commands and they appear to work on my oven. I also added a note about setpoint variable. The oven setpoint will not truly update until the ```run()``` command is sent. Changing setpoint and reading back setpoint before sending the ```run()``` command will result in the setpoint not being changed. Additionally, while the oven is running, if the setpoint is changed, the oven will not ramp to the new temperature until the ```run()``` command is sent again. 

Please let me know if there is anything else I should tweak or needs clarification. Thanks!

msmttchr commented on pymeasure/instruments/thermotron/thermotron3800.py:
```diff
@@ -0,0 +1,161 @@
+    temperature = Instrument.measurement(
+        "PVAR1?", """ Reads the current temperature of the oven
+        via built in thermocouple
+
```
Is unit Celsius?

msmttchr replied:
Is unit Celsius?

msmttchr reviewed (approved):
Looks good to me. I just left a minor comment on the temperature unit.
Thanks for your contribution
